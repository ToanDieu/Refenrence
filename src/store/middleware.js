import { matchPath } from "react-router-dom";
import { CHANGE_PAGE, NOTIFY_BUTTON_CLICK } from "../constants/actionTypes";

const changeToBasePage = id => ({
  type: CHANGE_PAGE,
  payload: {
    pageType: "base",
    detail: { id }
  }
});

export const trackPageNavigate = ({
  dispatch
  // getState
}) => next => action => {
  if (action.type === "@@router/LOCATION_CHANGE") {
    console.log("trackViews__action: ", action);
    const pathName = action.payload.location.pathname;

    const isEnteringBaseCases = matchPath(pathName, {
      path: "/bases/:baseID/cases"
    });

    const isEnteringBaseDetail = matchPath(pathName, {
      path: "/bases/:baseID"
    });

    // console.log("isEnteringBaseDetail", isEnteringBaseDetail, action);

    const isEnteringBaseWorkflow = matchPath(pathName, {
      path: "/bases/:baseID/workflows/*/actions"
    });

    const isEnteringNotification = matchPath(pathName, {
      path: "/notifications"
    });

    if (isEnteringNotification) {
      dispatch({
        type: NOTIFY_BUTTON_CLICK
      });
    }

    if (isEnteringBaseCases) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseCases.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }

    if (isEnteringBaseDetail) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseDetail.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }

    if (isEnteringBaseWorkflow) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseWorkflow.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }
  }

  return next(action);
};
