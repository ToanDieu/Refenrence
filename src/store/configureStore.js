import { createStore, compose, applyMiddleware } from "redux";
import reduxImmutableStateInvariant from "redux-immutable-state-invariant";
import thunk from "redux-thunk";
import { routerMiddleware } from "connected-react-router";

import history from "@/utils/history";
import createReducer from "../reducers";
import { trackPageNavigate } from "./middleware";

function configureStoreProd(initialState) {
  const reactRouterMiddleware = routerMiddleware(history);
  const middlewares = [thunk, reactRouterMiddleware, trackPageNavigate];

  const store = createStore(
    createReducer(),
    initialState,
    compose(applyMiddleware(...middlewares))
  );

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  return store;
}

function configureStoreDev(initialState) {
  const reactRouterMiddleware = routerMiddleware(history);
  const middlewares = [
    reduxImmutableStateInvariant(),
    thunk,
    reactRouterMiddleware,
    trackPageNavigate
  ];

  /* eslint-disable no-underscore-dangle */
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; // add support for Redux dev tools
  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(applyMiddleware(...middlewares))
  );

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept("../reducers", () => {
      const nextReducer = require("../reducers").default; // eslint-disable-line global-require
      store.replaceReducer(nextReducer);
    });
  }

  return store;
}

const store =
  process.env.NODE_ENV === "production"
    ? configureStoreProd()
    : configureStoreDev();

export { history, store };
