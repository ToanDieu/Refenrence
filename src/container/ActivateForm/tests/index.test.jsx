import React from "react";
import { render, fireEvent } from "react-testing-library";
import { Provider } from "react-redux";
import { initialize, addTranslationForLanguage } from "react-localize-redux";
import { Notification } from "element-react";

import languages from "@/i18n/languages.json";
import english from "@/i18n/en.json";
import { store as appStore } from "@/store/configureStore";
import { ActivateForm } from "../index";

describe("<ActivateForm />", () => {
  let store;
  const translate = jest.fn();

  beforeAll(() => {
    store = appStore;
    const defaultLanguage = "en";
    store.dispatch(initialize(languages, { defaultLanguage }));
    store.dispatch(addTranslationForLanguage(english, "en"));
  });

  it("should render the form", () => {
    const workflow = { name: "Test" };
    const { container } = render(
      <Provider store={store}>
        <ActivateForm translate={translate} workflow={workflow} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should handles when activate button was clicked", () => {
    const workflow = { name: "Test" };
    const clickSpy = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <ActivateForm
          translate={translate}
          workflow={workflow}
          setDialogVisible={clickSpy}
        />
      </Provider>
    );
    fireEvent.click(container.lastChild);
    expect(clickSpy).toBeCalledWith(true);
  });

  it("should handles when submit button was clicked", () => {
    const workflow = { name: "Test" };
    const submitSpy = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <ActivateForm
          translate={translate}
          workflow={workflow}
          activateWorkflow={submitSpy}
        />
      </Provider>
    );
    fireEvent.click(container.querySelector("button[type=submit]"));
    expect(submitSpy).toBeCalledTimes(1);
  });

  it("should handles when dialog was closed", () => {
    const workflow = { name: "Test" };
    const closeSpy = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <ActivateForm
          translate={translate}
          workflow={workflow}
          setDialogVisible={closeSpy}
        />
      </Provider>
    );
    const closeButton = container.querySelector("button.el-dialog__headerbtn");
    fireEvent.click(closeButton);
    expect(closeSpy).toBeCalledWith(false);
  });

  it("should show error notification when re-render within error", () => {
    const workflow = { name: "Test" };
    const spy = jest.spyOn(Notification, "error");
    render(
      <Provider store={store}>
        <ActivateForm
          translate={translate}
          workflow={workflow}
          error={{ message: "Error" }}
        />
      </Provider>
    );
    expect(spy).toHaveBeenCalled();
  });

  it("should show success notification when re-render within activated", () => {
    const workflow = { name: "Test" };
    const spy = jest.spyOn(Notification, "error");
    render(
      <Provider store={store}>
        <ActivateForm translate={translate} workflow={workflow} activated />
      </Provider>
    );
    expect(spy).toHaveBeenCalled();
  });
});
