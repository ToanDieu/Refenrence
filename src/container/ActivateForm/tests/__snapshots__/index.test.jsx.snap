// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActivateForm /> should render the form 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 .tseicon {
  font-size: 40px;
  margin-right: 20px;
}

.emotion-0 ul {
  list-style: initial;
  padding-left: 20px;
}

.emotion-0 ul li {
  margin-bottom: 20px;
}

.emotion-0 p {
  padding-left: 20px;
}

.emotion-1 {
  text-transform: uppercase;
}

.emotion-3 {
  vertical-align: top;
  margin-right: 12px;
  height: 40px;
  padding-left: 15px;
  padding-right: 15px;
}

.emotion-3 .tseicon {
  font-size: 24px;
  vertical-align: middle;
  line-height: 14px;
  margin-right: 7px;
  opacity: 0.5;
}

<div>
  <div>
    <div
      class="el-dialog__wrapper"
      style="z-index: 1013; display: none;"
    >
      <div
        class="el-dialog el-dialog--small"
        style="top: 15%;"
      >
        <div
          class="el-dialog__header"
        >
          <span
            class="el-dialog__title"
          />
          <button
            class="el-dialog__headerbtn"
            type="button"
          >
            <i
              class="el-dialog__close el-icon el-icon-close"
            />
          </button>
        </div>
        <div
          class="el-dialog__body"
        >
          <div
            class="emotion-0"
          >
            <i
              class=" tseicon tseicon-warning"
              css="[object Object]"
            />
            <div>
              <ul>
                <li>
                  You won’t be able to add triggers or actions when the automation is active.
                </li>
                <li>
                  You won’t be able to deactivate the workflow
                </li>
              </ul>
              <p>
                Do you want to activate the automation?
              </p>
            </div>
          </div>
        </div>
        <div
          class="el-dialog__footer dialog-footer"
        >
          <button
            class="el-button el-button--default emotion-1"
            type="button"
          >
            <span />
          </button>
          <button
            class="el-button el-button--primary emotion-1"
            type="submit"
          >
            <span />
          </button>
        </div>
      </div>
    </div>
    <div
      class="v-modal"
      style="z-index: 1012; display: none;"
    />
  </div>
  <button
    class="el-button el-button--primary emotion-3"
    type="button"
  >
    <span />
  </button>
</div>
`;
